### PR TITLE
TDD live tests support for Swift

### PR DESCRIPTION
**Pull request by polac24**
Hi,
this pull-request contains a feature to automatically execute all tests covering implementation file (very useful in TDD). It works only for Swift development.

Thanks to this feature, injectionforxcode provides a great TDD developer experience, where tests are executed every time, developer saves file with ⌘+S. I found is very useful.

**It contains:**
* little refactor of `injectSource.pl` to scan entire log in order to recognize if there is any unit test module (with XCTest)
* build injected files without "Code coverage" flag
* unit tests-related logic is placed in new package `InjectUnitTests.pm`
* when XCTest has been found, look for all Swift tests that directly reference at least one symbol from our injecting file
* rebuild swift project (`swiftc` with -incremental flag) and update entire module map
* rebuild testing project (`swiftc` with -incremental flag)
* calling `swiftc` is required to parse its JSON'ish response and learn commands for unit tests
* in case, that we accidentally learnt command to compile `$selectedFile` when doing unit-test dance (and standard way did not learnt it), use it (optimistic `$learnt`) 
* for unit tests, get rid of `-filelist` and replace it with inline list of files (list is taken from `swiftc` command)
* inject unit tests into a bundle
* link all resources (symbolic link) into injection bundle (it costs us nothing, but gives a possibility to call in tests `Bundle(for: MyTests.self).path(forResource: ....`
* update readme.md
* add an animation to readme file

**Requirements**
This pull-request assumes that user did integrate `InjectionTDD` framework (https://github.com/polac24/InjectionTDD). *If a developer wants to see a notification readme there shows how to add this breakpoint to Xcode.*

**Impact on existing cases**
In general, users that did not use ⌘+U (framework `XCTest` has not be found in log), should not observer any significant overhead (scanning even huge log is really quick, comparing to xcodebuild). 

**Tested and verified environment**
* Swift3.1 and Swift4
* Xcode 8.3.2 and Xcode 9.0
* variety of iOS projects: from small one-file (just to make POC), my internal medium-size to huge open-source project from [kickstarter](https://github.com/kickstarter/ios-oss)

### CODE REVIEW

**Comment by johnno1962** [comment]
Hi Bartosz, Thanks for looking into this for me but I’m a little reluctant to make any changes to injectionforxcode at the moment as it seems to be running fine for most users and is rather brittle. I’ve added a note to the injection README pointing users to your REPO if they would like to do TDD. (P.S. it looks like you’ll need to add “InjectUnitTests.pm” to the project so it runs correctly.)

**Comment by johnno1962** [comment]
On second thoughts, talking to @orta I’ll merge for now If you squash it up so I can revert easily if there are any problems. It seems to work once you fix the missing library file. Thanks for this 👍

**Comment by orta** [comment]
Amazing PR, I've [tried to do this before](https://github.com/orta/cocoapods-xcautotest/) and failed. Your user-land code is also much more elegant - great job.

@polac24 is it also possible that you can add some way to allow running the full test suite also? E.g. either opting in, or opting out of listening? 

I'd imagine in a large codebase, you'd want to be able to provide an scheme for watch mode and a scheme for running the full tests. For example:

screen shot 2017-09-24 at 7 23 23 am

File: InjectionPluginLite/injectSource.pl [hunk 1]
```diff
@@ -377,6 +417,14 @@ sub mtime {
         or die "Could not locate object file in: $learnt";
     ###$learnt =~ s/( -DDEBUG\S* )/$1-DINJECTION_BUNDLE /;
 
+    # Disable Code coverage for injection file
+    # swift
+    $learnt =~ s/\-profile\-generate//g;
```
Comment (johnno1962) [inline InjectionPluginLite/injectSource.pl]: Just curious, why do you escape the -’s?

**Comment by polac24** [comment]
@johnno1962, thanks for pointing out missing `InjectUnitTests.pm` in pbxproj. I added it and squashed PR.

@orta, it would be possible to run all the tests, but I was afraid that it can take too long to recompile all implementation and test files for the sake of running full suite of tests. Attaching all the implementation and test binaries every time into InjectionBundle is required due to static dispatch that Swift uses. Otherwise, **old** version of a struct/enum/extension could be used in our tests...
However, we can force developers to disable "Code coverage" for this schema and then reuse all ".o" binaries directly from incremental `swiftc` command.  Do you think that this requirement/limitation is reasonable? 
Thanks @orta, for your suggestion, I will definitely consider it. It totally makes sense for me!

File: InjectionPluginLite/injectSource.pl (hunk 1 above)
Comment (polac24) [reply InjectionPluginLite/injectSource.pl]: My mistake. I will update PR with just `-`

**Comment by johnno1962** [comment]
Let me know when you’re done and I’ll perform some checks. I’ll merge this evening in case you have any last minute changes. Thanks again.

**Comment by polac24** [comment]
I'm done.

Thanks!

**Comment by johnno1962** [comment]
Do you have an example project with a test I can clone?

**Comment by polac24** [comment]
I added some samples to https://github.com/polac24/InjectionTDD/tree/master/Example

**Pull request merged**